### PR TITLE
Adding WriterBuilder::begin_2d

### PR DESCRIPTION
**Pull request by Yomguithereal**
Hi @ExpHP, here is the PR adding `WriterBuilder::begin_2d` as discussed in #78.

The only thing which remains unclear to me is whether shape should be written the other way around when using FORTRAN order, but I think not.

I have used `record_size` as per your suggestion.

Have a great day,

### CODE REVIEW

**Comment by ExpHP** [comment]
Sounds exciting!  I've got a bit of a headache right now but I will take a look this week

**Comment by ExpHP** [comment]
Taking a look right now

File: src/write.rs [hunk 1]
```diff
@@ -375,7 +396,9 @@ pub struct NpyWriter<Row: Serialize + ?Sized, W: Write> {
 
 enum ShapeInfo {
     // No shape was written; we'll return to write a 1D shape on `finish()`.
-    Automatic { offset_in_header_text: u64 },
+    Automatic1D { offset_in_header_text: u64 },
+    // Only partial shape was written; we'll return to write the missing part of shape on `finish()`.
+    Automatic2D { offset_in_header_text: u64, expected_record_size: u64 },
```
Comment (ExpHP) [inline src/write.rs]: I would suggest maybe `record_len`?  `expected_` doesn't seem appropriate as this is the field that *defines* the record size.  And I think a `_size` right next to an offset could be misunderstood as a size in bytes.  (at least, more likely than a `_len`)

File: src/write.rs [hunk 2]
```diff
@@ -467,6 +490,24 @@ impl<Row: Serialize + ?Sized , W: Write> NpyWriter<Row, W> {
...
+                }
+
+                // Write the size to the header
+                let shape_pos = self.start_pos.unwrap() + self.version_props.bytes_before_text() as u64 + offset_in_header_text;
+                let end_pos = self.fw.seek(SeekFrom::Current(0))?;
+
+                self.fw.seek(SeekFrom::Start(shape_pos))?;
+                let length = format!("{}", self.num_items / expected_record_size);
+                self.fw.write_all(length.as_bytes())?;
+                write!(self.fw, ", {}), }}", expected_record_size).unwrap();
```
Comment (ExpHP) [inline src/write.rs]: I'm pretty sure we do need to flip the order of the axes here for Fortran.  The intention of the record len is for it to be the fast axis length.  The fast axis is in a different location between Fortran and C.

If you make arrays all aliased to share the same allocation in memory, the one with order "F" and the same shape as C will look like nonsense, while the one with swapped order *does* make sense (it is the transpose of `c`).
```python
import numpy as np
base = np.array([11, 12, 13, 21, 22, 23], dtype=np.int32)
c     = np.ndarray(shape=(2,3), dtype=np.int32, buffer=base, order="C")
fsame = np.ndarray(shape=(2,3), dtype=np.int32, buffer=base, order="F")
fswap = np.ndarray(shape=(3,2), dtype=np.int32, buffer=base, order="F")

print(f"c: (shape: {c.shape}\n{c}")
print(f"fsame: (shape: {fsame.shape})\n{fsame}")
print(f"fswap: (shape: {fswap.shape})\n{fswap}")

print("c == fsame  :", np.all(c == fsame))
print("c == fswap.T:", np.all(c == fswap.T))
print("same pointer:", c.ctypes.data == fswap.ctypes.data)
```

```
c: (shape: (2, 3)
[[11 12 13]
 [21 22 23]]
fsame: (shape: (2, 3))
[[11 13 22]           <-- 11, 13, 22? Nonsense grouping...
 [12 21 23]]
fswap: (shape: (3, 2))
[[11 21]
 [12 22]
 [13 23]]
c == fsame  : False
c == fswap.T: True
same pointer: True    <-- therefore DEFINITELY the same bytes
```

**Comment by ExpHP** [comment]
I really appreciate the division into two commits.  The first commit deserves a chef's kiss.

It took me a surprisingly long time to come up with compelling evidence for the transpose argument without needing to do some jank like "`np.save` and compare the written bytes after slicing the header away".  Surprisingly, both `bytes(array.data)` and `.tobytes(order="K")` end up allocating a second temporary buffer to rewrite the data into C contiguous order, producing different results from the payload written to an npy file. (The order="K" one is a [still-open bug](https://github.com/numpy/numpy/issues/16569) that was reported in 2020...).

I ended up needing to go as far as having 3 arrays all alias the exact same buffer in memory.   While I can't directly show that they share the same bytes, I can show that they share the same address in memory! :joy:

**Comment by ExpHP** [comment]
* [ ] Note to self that the definition of `FILLER_FOR_UNKNOWN_SIZE` is not consistent with its usage, as it does not include the closing parens.

File: src/write.rs (hunk 1 above)
Comment (Yomguithereal) [reply src/write.rs]: Should I also rename the argument to `WriterBuilder::begin_2d` from `record_size` to `record_len` also, then?

**Comment by Yomguithereal** [comment]
Whay is `FILLER_FOR_UNKNOWN_SIZE` is 19 by the way? Shouldn't it be 20, e.g. the maximum number of digits a u64 can have (without even considering your note above considering the closing parens)?

File: src/write.rs (hunk 2 above)
Comment (Yomguithereal) [reply src/write.rs]: I'll deal with the shape swap then. I guess this mean the `NpyWriter` will have to track its `Order` then. Is this alright with you? Or should I keep it in the `ShapeInfo::Automatic2D` variant instead?

**Comment by ExpHP** [comment]
I was wondering that myself.  I don't recall how I chose that number.  I should add a unit test for begin_1d that has item size zero and len u64::MAX.

File: src/write.rs (hunk 2 above)
Comment (ExpHP) [reply src/write.rs]: I think you could put it either on ShapeHint::Automatic2D or NpyWriter.  I have no problem with either, it's private either way.

File: src/write.rs (hunk 1 above)
Comment (ExpHP) [reply src/write.rs]: Yes let's do that.

**Comment by Yomguithereal** [comment]
> I was wondering that myself. I don't recall how I chose that number. I should add a unit test for begin_1d that has item size zero and len u64::MAX.

I can investigate and add a unit test for that while I am amending the PR if you want.

**Comment by ExpHP** [comment]
Sure, I'd appreciate it.  Now that I think about it I probably have one already (can't check on my phone), but it might be succeeding too easily due to the extra space for alignment... that could make it really hard to make a test that reliably detects a regression and might not be worth it.

As an alternative, we could maybe just write the max `u64` value as a placeholder instead of a `const` str.  That should be easy to reason about, if we pull out a function for writing a shape for a given len that gets used both at begin and finish. (so that there can be no mismatch in length of the preceeding or trailing characters)

**Comment by Yomguithereal** [comment]
> As an alternative, we could maybe just write the max u64 value as a placeholder instead of a const str. 

I think using your `****...` placeholder is safer. I don't like the idea of using the max `u64` there (apart for testing purposes). Else some parsers/readers might pre-allocate the data to a max `u64` and crash before even realizing the file is lying to them. At least with the `*` it will crash before :)

**Comment by Yomguithereal** [comment]
@ExpHP I have amended the PR to:

1. rename `record_size` to `record_len`
2. Bump `FILLER_FOR_UNKNOWN_SIZE` to 20 (max number of digits in a u64)
3. added correct `Order::Fortran` support (and removed shameful bad `.unwrap()` calls in `finish_`)

I could not find what you think is problematic currently regarding `FILLER_FOR_UNKNOWN_SIZE`. It seems to do its job correctly and is not supposed to take the `",), }"` part into account in its size. Maybe it's just a question of amending the comment above its definition?

**Comment by ExpHP** [comment]
> At least with the * it will crash before :)

Good argument, I agree.

> Maybe it's just a question of amending the comment above its definition?

Eh, it's just the fact that the total number of characters written depends on the number of spaces and presence/lack of trailing commas, and that it's not immediately obvious that these factors match between the two places.

I will merge this as is, play around with it a bit and prep for release.

**Comment by ExpHP** [comment]
also going to audit for any other `unwrap`s on IO ~~and look into how this one slipped through~~ ...oh, it was yours.  Gotcha.  Still gonna do an audit tho
